In [ ]:
import pandas as pd

# Starting data sets

In [ ]:
#The following are copies of the originally provided raw data from WBS
# orders.csv
orders = "https://drive.google.com/file/d/1gyoqm_HkCXYxPEAotoHy3AnjQb062db0/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+orders.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
orderlines = "https://drive.google.com/file/d/1jXT7JdxARmUrVc4qqE1wUPTuaEw7tjxw/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+orderlines.split("/")[-2]
orderlines = pd.read_csv(path)

# products.csv
products = "https://drive.google.com/file/d/1HUKPnuS9RJjrTstonYmTHDYGZ0foZbz9/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+products.split("/")[-2]
products = pd.read_csv(path)

# brands.csv
brands = 'https://drive.google.com/file/d/1C9fPQuinXOYkWe0NzcDgG3vLDg2seV1b/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+brands.split("/")[-2]
brands = pd.read_csv(path)

**Starting data status**

orders: (226909, 4)

orderlines:(293983, 7)

products:(19326, 7)

brands:(187, 2)

# Manipulation of data sets

### Removal of duplicates

In [ ]:
#creating copies
orders_df = orders.copy()

orderlines_df = orderlines.copy()

products_df=products.copy()

brands_df=brands.copy()

In [ ]:
# orders_df - no duplicates
orders_df.duplicated().sum()

0

In [ ]:
# orderlines_df - no duplicates
orderlines_df.duplicated().sum()

0

In [ ]:
#a very large amount of products are duplicated (8746, or 45%)
products_df.duplicated().sum()

8746

In [ ]:
#dropping duplicates
products_df.drop_duplicates(inplace=True)

In [ ]:
#verifying duplicates have been removed
#original number 19326 - -8746 removed = updated number 10580
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


In [ ]:
#brands_df - no duplicates
brands_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   short   187 non-null    object
 1   long    187 non-null    object
dtypes: object(2)
memory usage: 3.0+ KB


**Data status after duplicates have been removed**

orders: (226909, 4) => (226909, 4)

orderlines:(293983, 7) ==> (293983, 7)

products:(19326, 7) ==> (10580, 7)

brands:(187, 2)==> (187, 2)

### Removal of missing values

In [ ]:
#orders missing 5 values
orders_df.total_paid.isna().sum()

5

In [ ]:
#orders missing values removed
orders_df = orders_df.loc[~orders_df.total_paid.isna(), :]

In [ ]:
#orderlines has no missing values
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [ ]:
#products has 3 columns missing values
# 7 values are missing from the desc column
# 46 values are missing from the price column
# 50 values are missing from the type column -- !!!BUT we will not remove it as it was not done officially by WBS!!!
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


In [ ]:
#copying product names to the missing description values for the same row, in case we want to use the description column at some point (per instructor)
#the 7 rows that previously had missing information for "desc" have not been deleted, because of the above copying step

products_df.loc[products_df['desc'].isna(), 'desc'] = products_df.loc[products_df['desc'].isna(), 'name']

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10580 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


In [ ]:
#deleting price rows with missing values
products_df = products_df.loc[~products_df.price.isna(), :]

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10534 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10534 non-null  object
 1   name         10534 non-null  object
 2   desc         10534 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10534 non-null  object
 5   in_stock     10534 non-null  int64 
 6   type         10484 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.4+ KB


**Data status after missing values have been removed**

orders: (226909, 4) => (226909, 4) ==> (226904, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7)

brands:(187, 2)==> (187, 2) ==> (187, 2)

### Removal of two-dot problem (and column "promo_price")

Removal from orderlines DataFrame

In [ ]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

In [ ]:
orderlines_df.shape

(216250, 7)

In [ ]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

Removals from products DataFrame

In [ ]:
#looking at how many rows in products (column "price") have the two-dot problem and have corrupted entry (i.e. multiple figures after )= 542
price_problems_number = products_df.loc[(products_df.price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
price_problems_number

542

In [ ]:
#removing the problem from products
products_df = products_df.loc[(~products_df.price.astype(str).str.contains("\d+\.\d+\.\d+"))&(~products_df.price.astype(str).str.contains("\d+\.\d{3,}")), :]

In [ ]:
#original count was 10580 - 588 removed = 9992 now
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9992 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          9992 non-null   object
 1   name         9992 non-null   object
 2   desc         9992 non-null   object
 3   price        9992 non-null   object
 4   promo_price  9992 non-null   object
 5   in_stock     9992 non-null   int64 
 6   type         9946 non-null   object
dtypes: int64(1), object(6)
memory usage: 624.5+ KB


In [ ]:
#looking at how many rows in products (column "promo_price") have the two-dot problem  = 9232
promo_problems_number = products_df.loc[(products_df.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
promo_problems_number

9232

In [ ]:
#over 90% of the values in promo_price are two-dotted
print(f"The column promo_price has in total {promo_problems_number} wrong values. This is {round(((promo_problems_number / products_df.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9232 wrong values. This is 92.39% of the rows of the DataFrame


In [ ]:
#as over 90% of the data in the column is corrupt, complete column will be deleted

#completely deleting promo_price column
products_df = products_df.drop(columns=["promo_price"])

**Data status after two-dot problem has been removed**

orders: (226909, 4) => (226909, 4) ==> (226904, 4) ==> (226904, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7) ==> (216250, 7)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7) ==> (9992, 6)

brands:(187, 2)==> (187, 2) ==> (187, 2) ==> (187, 2)

### Removal of unwanted states

In [ ]:
#making copies
o_df=orders_df.copy()
ol_df=orderlines_df.copy()
p_df=products_df.copy()
b_df=brands_df.copy()

In [ ]:
#changing datatype
o_df["created_date"]=pd.to_datetime(o_df["created_date"])
ol_df["date"]=pd.to_datetime(ol_df["date"])

In [ ]:
#checking count per 'state':

#7233 rows are "cancelled" only
#o_df[o_df['state'].str.contains('cancelled', case=False, na=False)]

#117809 rows are "shopping basket" only
#o_df[o_df['state'].str.contains('shopping basket', case=False, na=False)]

#14374 rows are "pending" only
#o_df[o_df['state'].str.contains('pending', case=False, na=False)]

#40883 rows are "place order" only
#o_df[o_df['state'].str.contains('place order', case=False, na=False)]

#46605 rows are "completed" only
#o_df[o_df['state'].str.contains('completed', case=False, na=False)]

order_id         created_date  total_paid      state
1         241423  2017-11-06 13:10:02      136.15  Completed
2         242832  2017-12-31 17:40:03       15.76  Completed
3         243330  2017-02-16 10:59:38       84.98  Completed
5         245275  2017-06-28 11:35:37      149.00  Completed
6         245595  2017-01-21 12:52:47      112.97  Completed
...          ...                  ...         ...        ...
226549    527042  2018-03-14 11:47:50       18.98  Completed
226577    527070  2018-03-14 11:50:48       24.97  Completed
226581    527074  2018-03-14 11:51:42       24.97  Completed
226603    527096  2018-03-14 11:58:40       34.96  Completed
226619    527112  2018-03-14 12:03:52       14.98  Completed

[46605 rows x 4 columns]

In [ ]:
#keeping only rows with value "Completed" for column "state"
#original number of rows before filtering: 226904
#updated number of rows after filtering: 46605

o_df=o_df.query("state=='Completed'")

In [ ]:
o_df.shape

(46605, 4)

**Data status after removal of unwanted states**

orders: (226909, 4) => (226909, 4) ==> (226904, 4) ==> (226904, 4) ==> (46605, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7) ==> (216250, 7) ==> (216250, 7)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7) ==> (9992, 6) ==> (9992, 6)

brands:(187, 2)==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2)

### Keeping only orders present in both orders and orderlines

In [ ]:
orders_orderlines = o_df.merge(ol_df,
                left_on='order_id',
                right_on='id_order',
                how='inner')

In [ ]:
#56714=original number of rows before removing order_id rows missing from the products table
orders_orderlines.shape

(216057, 11)

In [ ]:
order_ids = orders_orderlines['order_id'].unique()

In [ ]:
# keeping only the common order_ids in orders
o_df = o_df.loc[o_df["order_id"].isin(order_ids), :]

# keeping only the common order_ids in orderlines
ol_df = ol_df.loc[ol_df["id_order"].isin(order_ids), :]

In [ ]:
#orders and orderlines now have same amount of unique order_ids
o_df["order_id"].nunique(), ol_df["id_order"].nunique()

(170069, 170069)

**Data status after keeping only orders present in both orders and orderlines**

orders: (226909, 4) => (226909, 4) ==> (226904, 4) ==> (226904, 4) ==> (46605, 4) ==> (43064, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7) ==> (216250, 7) ==> (216250, 7) ==> (56714, 7)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7) ==> (9992, 6) ==> (9992, 6) ==> (9992, 6)

brands:(187, 2)==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2)

### Removal of unknown products

In [ ]:
orderlines_products = ol_df.merge(p_df, how="left", on = "sku")[["id_order","sku","name" ]]
orderlines_products.sample(10)

id_order        sku                                               name
92117     390730    OWC0218  SSD expansion kit OWC Aura Pro 6G 120GB MacBoo...
93198     391819    TPL0036                                                NaN
114051    413539    APP1155  Case Apple iPhone 6 Plus / 6s Plus Silicone Ca...
186849    494583    PHI0076                   Philips Hue Motion Motion Sensor
190724    498974    SYN0093                  Synology Pack 8 IP camera license
34558     333106    SAN0098  SanDisk Connect Wireless Flash Drive 32GB Wire...
91497     390092    IFR0013   GlassGuard iFrogz iPhone Screen Saver 7 / 6s / 6
24518     323388  FCM0016-2    Mac memory FCM 8GB (2x4GB) SO-DIMM DDR3 1333MHz
74789     373022    APP0409  Apple 85W MagSafe 2 charger MacBook Pro Retina...
162330    466658    SEA0110         Seagate Hard Drive 4TB NAS IronWolf SATA 3

In [ ]:
orders_to_delete = orderlines_products.loc[orderlines_products.name.isna(), "id_order"].unique()
orders_to_delete

array([299576, 299638, 299695, ..., 527170, 527191, 527367])

In [ ]:
#keeping only the orders in orders and orderlines that are not corrupted

In [ ]:
o_df = o_df.loc[~o_df.order_id.isin(orders_to_delete),:]

In [ ]:
ol_df = ol_df.loc[~ol_df.id_order.isin(orders_to_delete),:]

In [ ]:
o_df["order_id"].nunique(), ol_df["id_order"].nunique()

(163750, 163750)

**Data status after removal of rows where products were unknown**

orders: (226909, 4) => (226909, 4) ==> (226904, 4) ==> (226904, 4) ==> (46605, 4) ==> (43064, 4) ==> (41701, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7) ==> (216250, 7) ==> (216250, 7) ==> (56714, 7) ==> (54324, 7)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7) ==> (9992, 6) ==> (9992, 6) ==> (9992, 6) ==>  (9992, 6)

brands:(187, 2)==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2)

### Removal of outliers

In [ ]:
#creating a new column with values of the other two tables multiplied
ol_df.loc[:,'unit_price_total'] = ol_df['unit_price'] * ol_df['product_quantity']

In [ ]:
ol_agg = ol_df.groupby("id_order", as_index=False)["unit_price_total"].sum()

In [ ]:
ol_agg_o = ol_agg.merge(o_df, left_on="id_order", right_on="order_id")

In [ ]:
ol_agg_o_dif = ol_agg_o.copy()

In [ ]:
#difference between total_paid and unit_price_total
ol_agg_o_dif["order_price_dif"] = ol_agg_o["total_paid"] - ol_agg_o["unit_price_total"]

In [ ]:
# Removing outliers from ol_agg_o_diff (orderlines aggregated and merged with orders, dif between total_price and unit_price_total)

# Calculating Q1 (25th percentile) and Q3 (75th percentile)
Q1 = ol_agg_o_dif["order_price_dif"].quantile(0.25)
Q3 = ol_agg_o_dif["order_price_dif"].quantile(0.75)
IQR = Q3 - Q1

# Defining the outlier range
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filtering the DataFrame to remove outliers
ol_agg_o_dif_no_outliers = ol_agg_o_dif[(ol_agg_o_dif["order_price_dif"] >= lower_bound) & (ol_agg_o_dif["order_price_dif"] <= upper_bound)]

In [ ]:
#Removing the outliers' unique order_id/id_order from orders as well as from orderlines_products merged DataFrame.
normal_diff_list = ol_agg_o_dif_no_outliers["id_order"]

In [ ]:
o_df = o_df.loc[o_df["order_id"].isin(normal_diff_list),:]

In [ ]:
ol_df = ol_df.loc[ol_df["id_order"].isin(normal_diff_list),:]

In [ ]:
#40985, 40985
o_df["order_id"].nunique(), ol_df["id_order"].nunique()

(161088, 161088)

**Data status after removal of outliers**

orders: (226909, 4) => (226909, 4) ==> (226904, 4) ==> (226904, 4) ==> (46605, 4) ==> (43064, 4) ==> (41701, 4) ==> (40985, 4)

orderlines:(293983, 7) ==> (293983, 7) ==> (293983, 7) ==> (216250, 7) ==> (216250, 7) ==> (56714, 7) ==> (54324, 7) ==> (53231, 8)

products:(19326, 7) ==> (10580, 7) ==> (10534, 7) ==> (9992, 6) ==> (9992, 6) ==> (9992, 6) ==>  (9992, 6) ==> (9992, 6)

brands:(187, 2)==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2) ==> (187, 2)

In [ ]:
o_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 161088 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      161088 non-null  int64         
 1   created_date  161088 non-null  datetime64[ns]
 2   total_paid    161088 non-null  float64       
 3   state         161088 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 6.1+ MB


# Cleaned (and categorized) data sets

In [ ]:
# orders_clean_df.csv
orders_clean_df = "https://drive.google.com/file/d/1ZEuWF3-pKYC-i6fP8unAAG_JMqU2KEyc/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+orders_clean_df.split("/")[-2]
orders_clean_df = pd.read_csv(path)

# orderlines_clean_df.csv
orderlines_clean_df = "https://drive.google.com/file/d/1IbRAfLGhL7Yf_b8CVkknwhnzsdTBeCIt/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+orderlines_clean_df.split("/")[-2]
orderlines_clean_df = pd.read_csv(path)

# products_categorized_df.csv
products_categorized_df = "https://drive.google.com/file/d/12O5WsGIjbpwmKb-yU7ovn_UUNntvFOpa/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+products_categorized_df.split("/")[-2]
products_categorized_df = pd.read_csv(path)

# brands_clean_df.csv
brands_clean_df = "https://drive.google.com/file/d/1V33HMpSmSLjWSzmGv4is26QKIC_cqNpk/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+brands_clean_df.split("/")[-2]
brands_clean_df = pd.read_csv(path)